tools/hab-channel-list-pkgs/hab-channel-list-pkgs.sh

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+#
+# Returns a list of all fully qualified package identifiers that exist in a
+# given release channel. Prints the package identifiers to standard out, one
+# entry per line.
+#
+set -eu
+if [ -n "${DEBUG:-}" ]; then set -x; fi
+
+main() {
+  need_cmd curl
+  need_cmd jq
+
+  # Set the default Builder URL
+  : ${HAB_BLDR_URL:=https://bldr.habitat.sh}
+  # Channel name is required
+  channel="${1:-}"
+  if [[ -z "$channel" ]]; then
+    echo "Usage: $(basename $0) <CHANNEL>" >&2
+    exit 1
+  fi
+  # Calculate the channel URL from which to fetch
+  url="$HAB_BLDR_URL/v1/depot/channels/core/$channel/pkgs"
+
+  # Determine the number of results per page
+  npp="$(curl -s "$url" | jq '.range_end - .range_start + 1')"
+
+  # Determine the number of pages to fetch. It sort of sucks that jq's `ceil`
+  # doesn't work...going to do this the hard way...
+  num_pages=$(curl -s "$url" \
+    | jq "if ((.total_count / $npp) - ((.total_count / $npp) | floor)) > 0 then
+          ((.total_count / $npp) | floor) + 1
+        else
+          ((.total_count / $npp) | floor)
+        end")
+
+  # Fetch each page and output a fully qualified package identifier, one per
+  # line
+  for (( n=0; n<$num_pages; n+=1)) do
+    curl -s "${url}?range=$(( ${n}*${npp} ))" \
+      | jq --raw-output '.data[] | "\(.origin)/\(.name)/\(.version)/\(.release)"'
+  done
+}
+
+need_cmd() {
+  if ! command -v "$1" > /dev/null 2>&1; then
+    echo "Required command '$1' not found on PATH" >&2
+    exit 127
+  fi
+}
```

Comment:
I wish it were easier to share libraries of common Bash code... I just keep carting this function around everywhere I need it 😦

Reply:
Yeah, we could make more of an effort here, I have some ideas, assuming that they end up as Habitat packages.

However, in the spirit of appropriating all good things, I took this from the pretty fantastic [rustup-init.sh](https://github.com/rust-lang-nursery/rustup.rs/blob/80f9ec72490717132c6c70eb00ebdc98c2520d9c/rustup-init.sh#L332-L336) which is used to install the rustup binary. It was reading through this script that I realized making a `main()` function at the top, and calling it at the bottom was very………rusty almost.